spark/src/main/scala/geotrellis/spark/io/s3/S3LayerReader.scala

```diff
@@ -0,0 +1,67 @@
+package geotrellis.spark.io.s3
+
+import com.typesafe.scalalogging.slf4j.LazyLogging
+import geotrellis.spark._
+import geotrellis.spark.io._
+import geotrellis.spark.io.json._
+import geotrellis.spark.io.avro._
+import geotrellis.spark.io.index.KeyIndex
+import org.apache.avro.Schema
+import org.apache.spark.SparkContext
+import spray.json.{JsObject, JsonFormat}
+import spray.json.DefaultJsonProtocol._
+import AttributeStore.Fields
+
+import scala.reflect.ClassTag
+
+
+/**
+ * Handles reading raster RDDs and their metadata from S3.
+ *
+ * @param attributeStore  AttributeStore that contains metadata for corresponding LayerId
+ * @param getCache        Optional cache function to be used when reading S3 objects.
+ * @tparam K              Type of RDD Key (ex: SpatialKey)
+ * @tparam TileType       Type of RDD Value (ex: Tile or MultiBandTile )
+ * @tparam Container      Type of RDD Container that composes RDD and it's metadata (ex: RasterRDD or MultiBandRasterRDD)
+ */
+class S3LayerReader[K: Boundable: AvroRecordCodec: JsonFormat: ClassTag, TileType: AvroRecordCodec: ClassTag, Container[_]](
```

Comment:
Is it possible for the `Container` type to not specify type params (so the user would have to fully specify), and have object applys do some type translation so we get the generic machinery without sacrificing the cleaner API?

e.g.
```
      // This is the actual call to new
      val spaceTimeReader = new S3LayerReader[SpatialKey, Tile, RasterRDD[SpatialKey]](attributeStore){
        override val getS3Client = () => new MockS3Client
      }
  
     // This is an object apply that delegates to the above call
      val spaceTimeReader = S3LayerReader[SpatialKey, Tile, RasterRDD](attributeStore){
        override val getS3Client = () => new MockS3Client
      }
```